File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/1637-날카로운 눈/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 20005

int N, K;
int info[NM][3];

void input() {
	cin >> N;
	for (int i = 1; i <= N; i++) for (int j = 0; j < 3; j++) cin >> info[i][j];
}

int counting(int A, int C, int B, int X) {
	if (X < A) return 0;
	if (C < X) return (C - A) / B + 1;
	return (X - A) / B + 1;
}

bool determination(ll candidate) {
	ll sum = 0;
	for (int i = 1; i <= N; i++) {
		sum += counting(info[i][0], info[i][1], info[i][2], candidate);
	}
	return sum % 2 == 1;
}

void pro() {
	ll L = 1, R = 1ll << 31, ans = 0, ansCnt = 0;
	while (L <= R) {
		ll mid = (L + R) / 2;
		if (determination(mid)) {
			ans = mid;
			R = mid - 1;
		}
		else L = mid + 1;
	}
	if (ans == 0) {
		cout << "NOTHING";
	}
	else {
		for (int i = 1; i <= N; i++) {
			if (info[i][0] <= ans && ans <= info[i][1] && (ans - info[i][0]) % info[i][2] == 0) {
				ansCnt++;
			}
		}
		cout << ans << " " << ansCnt;
	}
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}

File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/1637-날카로운 눈/solution.py
import sys
si = sys.stdin.readline
n = int(si())
info = []
for i in range(n):
    info.append(list(map(int, si().split())))

def count(A, C, B, X):
    if X < A: return 0
    if C < X: return (C - A) // B + 1
    return (X - A) // B + 1

def determination(candidate):
    sum = 0
    for i in info:
        sum += count(i[0], i[1], i[2], candidate)
    return sum % 2 == 1

l, r, ans, ansCnt = 1, 1 << 31, 0, 0
while l <= r:
    mid = (l + r) // 2
    if determination(mid):
        ans = mid
        r = mid - 1
    else:
        l = mid + 1

if ans == 0:
    print("NOTHING")
else:
    for i in info:
        if i[0] <= ans and ans <= i[1] and (ans - i[0]) % i[2] == 0:
            ansCnt += 1
    print(ans, ansCnt)